internals/updates/docker.go

```diff
@@ -0,0 +1,59 @@
+package updates
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/client"
+	"github.com/portainer/agent"
+)
+
+type DockerUpdaterCleaner struct {
+}
+
+func NewDockerUpdaterCleaner() *DockerUpdaterCleaner {
+	return &DockerUpdaterCleaner{}
+}
+
+func (du *DockerUpdaterCleaner) Clean(ctx context.Context) error {
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithVersion(agent.SupportedDockerAPIVersion))
+	if err != nil {
+		return err
+	}
+	defer cli.Close()
+
+	foundRunningContainer := false
+	containers, err := cli.ContainerList(ctx, types.ContainerListOptions{All: true})
+	if err != nil {
+		return fmt.Errorf("failed to list containers: %s", err.Error())
+	}
+	for _, container := range containers {
+		_, ok := container.Labels["io.portainer.hideStack"]
+		if ok {
+			if container.State == "exited" {
+				err = cli.ContainerRemove(ctx, container.ID, types.ContainerRemoveOptions{Force: true})
+				if err != nil {
+					return fmt.Errorf("failed to remove container: %s", err.Error())
+				}
+
+				if container.NetworkSettings != nil {
+					for _, networkSetting := range container.NetworkSettings.Networks {
+						err = cli.NetworkRemove(ctx, networkSetting.NetworkID)
+						if err != nil {
+							return fmt.Errorf("failed to remove network: %s", err.Error())
+						}
+					}
+				}
+			} else if container.State == "running" {
+				foundRunningContainer = true
```

Comment:
why not just return the error here?

Reply:
In the event that there are multiple containers that match, it is necessary to evaluate all of them. While it is preferable to have only one container with the label `io.portainer.updater=true`, this solution is not problematic.